src/internal/git/push.go

```diff
@@ -37,55 +45,68 @@ func PushAllDirectories(localPath string) error {
 	for _, path := range paths {
 		basename := filepath.Base(path)
 		spinner.Updatef("Pushing git repo %s", basename)
-		if err := push(path, tunnelUrl, spinner); err != nil {
+
+		repo, err := prepRepoForPush(path, gitServerURL, gitServerInfo.PushUsername)
+		if err != nil {
+			message.Warnf("error when preping the repo for push.. %v", err)
+			return err
+		}
+
+		if err := push(repo, path, spinner); err != nil {
 			spinner.Warnf("Unable to push the git repo %s", basename)
 			return err
 		}
 
 		// Add the read-only user to this repo
-		repoPathSplit := strings.Split(path, "/")
-		repoNameWithGitTag := repoPathSplit[len(repoPathSplit)-1]
-		repoName := strings.Split(repoNameWithGitTag, ".git")[0]
-		err = addReadOnlyUserToRepo(tunnelUrl, repoName)
-		if err != nil {
-			message.Warnf("Unable to add the read-only user to the repo: %s\n", repoName)
-			return err
+		if gitServerInfo.InternalServer {
+			repoPathSplit := strings.Split(path, "/")
+			repoNameWithGitTag := repoPathSplit[len(repoPathSplit)-1]
+			repoName := strings.Split(repoNameWithGitTag, ".git")[0]
+			err = addReadOnlyUserToRepo(gitServerURL, repoName)
+			if err != nil {
+				message.Warnf("Unable to add the read-only user to the repo: %s\n", repoName)
+				return err
+			}
 		}
+
 	}
 
 	spinner.Success()
 	return nil
 }
 
-func push(localPath, tunnelUrl string, spinner *message.Spinner) error {
-
+func prepRepoForPush(localPath, tunnelUrl, username string) (*git.Repository, error) {
 	// Open the given repo
 	repo, err := git.PlainOpen(localPath)
 	if err != nil {
-		return fmt.Errorf("not a valid git repo or unable to open: %w", err)
+		return nil, fmt.Errorf("not a valid git repo or unable to open: %w", err)
 	}
 
 	// Get the upstream URL
 	remote, err := repo.Remote(onlineRemoteName)
 	if err != nil {
-		return fmt.Errorf("unable to find the git remote: %w", err)
-
+		return nil, fmt.Errorf("unable to find the git remote: %w", err)
 	}
+
 	remoteUrl := remote.Config().URLs[0]
-	targetUrl := transformURL(tunnelUrl, remoteUrl)
+	targetUrl := transformURL(tunnelUrl, remoteUrl, username)
 
 	_, err = repo.CreateRemote(&goConfig.RemoteConfig{
 		Name: offlineRemoteName,
 		URLs: []string{targetUrl},
 	})
-
 	if err != nil {
-		return fmt.Errorf("failed to create offline remote: %w", err)
+		return nil, fmt.Errorf("failed to create offline remote: %w", err)
 	}
 
+	return repo, nil
+}
+
+func push(repo *git.Repository, localPath string, spinner *message.Spinner) error {
+
 	gitCred := http.BasicAuth{
-		Username: config.ZarfGitPushUser,
-		Password: config.GetSecret(config.StateGitPush),
+		Username: config.GetState().GitServer.PushUsername,
+		Password: config.GetState().GitServer.PushPassword,
 	}
 
 	// Since we are pushing HEAD:refs/heads/master on deployment, leaving
```

Comment:
Does line 120 just work in Gitea land? (pushing to a repo that doesn't exist?)  I tested this with GitLab and it returned a `422` (full log below):

```
  ✔  Storing images in the zarf registry                                                                                       
  DEBUG   tunnel.Close()
└ (/home/wayne/workspace/zarf/src/internal/k8s/tunnel.go:254)
  DEBUG   Rewrite git URL: https://github.com/defenseunicorns/zarf.git -> https://gitlab.com/Racer159/mirror__github.com__defenseunicorns__zarf.git
└ (/home/wayne/workspace/zarf/src/internal/git/utils.go:49)

  DEBUG   Remove head copies for /tmp/zarf-3968008433/components/baseline/repos/mirror__github.com__defenseunicorns__zarf.git__v0.15.0
└ (/home/wayne/workspace/zarf/src/internal/git/utils.go:134)

  DEBUG   Remove git references /tmp/zarf-3968008433/components/baseline/repos/mirror__github.com__defenseunicorns__zarf.git__v0.15.0
└ (/home/wayne/workspace/zarf/src/internal/git/utils.go:162)

 WARNING  Unable to push the git repo mirror__github.com__defenseunicorns__zarf.git__v0.15.0                                   
  DEBUG   unable to push repo to the gitops service: unexpected client error: unexpected requesting "https://gitlab.com/Racer159/mirror__github.com__defenseunicorns__zarf.git/info/refs?service=git-receive-pack" status code: 422
└ (/home/wayne/workspace/zarf/src/internal/packager/deploy.go:280)
 WARNING  Unable to push repos to the Zarf Registry, retrying in 5 seconds...
  DEBUG   Rewrite git URL: https://github.com/defenseunicorns/zarf.git -> https://gitlab.com/Racer159/mirror__github.com__defenseunicorns__zarf.git
└ (/home/wayne/workspace/zarf/src/internal/git/utils.go:49)

 WARNING  error when preping the repo for push.. failed to create offline remote: remote already exists                        

  DEBUG   failed to create offline remote: remote already exists
└ (/home/wayne/workspace/zarf/src/internal/packager/deploy.go:280)
 WARNING  Unable to push repos to the Zarf Registry, retrying in 5 seconds...
  DEBUG   Rewrite git URL: https://github.com/defenseunicorns/zarf.git -> https://gitlab.com/Racer159/mirror__github.com__defenseunicorns__zarf.git
└ (/home/wayne/workspace/zarf/src/internal/git/utils.go:49)

 WARNING  error when preping the repo for push.. failed to create offline remote: remote already exists                        

  DEBUG   failed to create offline remote: remote already exists
└ (/home/wayne/workspace/zarf/src/internal/packager/deploy.go:280)
 WARNING  Unable to push repos to the Zarf Registry, retrying in 5 seconds...
  ✔  Zarf deployment complete
```

(Also these errors should likely be more than just warnings since the deployment was not actually complete)

Reply:
This seems to be the reason it failed:

`Could not create project: Path must not start or end with a special character and must not contain consecutive special characters.`

There are two underscores in a row in the project name.

Reply:
Once we resolve that we should have a note in the docs that this will only work with Gitea and GitLab (git push to non-existent repos is not a thing on Bitbucket or Azure DevOps Services (the other two big `git` hosting players).  Source Hut is up and coming and also does not support this.